Snowflake.Data.Tests/UnitTests/SFS3ClientTest.cs

```diff
@@ -0,0 +1,394 @@
+                        getObjectResponse.ContentLength = MockS3Client.ContentLength;
+                        getObjectResponse.Metadata.Add(SFS3Client.AMZ_IV, MockS3Client.AmzIV);
+                        getObjectResponse.Metadata.Add(SFS3Client.AMZ_KEY, MockS3Client.AmzKey);
+                        getObjectResponse.Metadata.Add(SFS3Client.AMZ_MATDESC, MockS3Client.AmzMatdesc);
+                        getObjectResponse.Metadata.Add(SFS3Client.SFC_DIGEST, MockS3Client.SfcDigest);
+
+                        return Task.FromResult(getObjectResponse);
+                    }
+                    else
+                    {
+                        throw MockS3Client.CreateMockAwsResponseError(request.BucketName, true);
+                    }
+                });
+            _client = new SFS3Client(_fileMetadata.stageInfo, MaxRetry, Parallel, _proxyCredentials, mockAmazonS3Client.Object);
+            _fileMetadata.client = _client;
+            _fileMetadata.stageInfo.location = requestKey;
+
+            // Act
+            FileHeader fileHeader = await _client.GetFileHeaderAsync(_fileMetadata, _cancellationToken).ConfigureAwait(false);
+
+            // Assert
+            AssertForGetFileHeaderTests(expectedResultStatus, fileHeader);
+        }
+
+        private void AssertForGetFileHeaderTests(ResultStatus expectedResultStatus, FileHeader fileHeader)
+        {
+            if (expectedResultStatus == ResultStatus.UPLOADED)
+            {
+                Assert.AreEqual(MockS3Client.ContentLength, fileHeader.contentLength);
+                Assert.AreEqual(MockS3Client.SfcDigest, fileHeader.digest);
+                Assert.AreEqual(MockS3Client.AmzIV, fileHeader.encryptionMetadata.iv);
+                Assert.AreEqual(MockS3Client.AmzKey, fileHeader.encryptionMetadata.key);
+                Assert.AreEqual(MockS3Client.AmzMatdesc, fileHeader.encryptionMetadata.matDesc);
+            }
+            else
+            {
+                Assert.IsNull(fileHeader);
+            }
+
+            Assert.AreEqual(expectedResultStatus.ToString(), _fileMetadata.resultStatus);
+        }
+
+        [Test]
+        [TestCase(MockS3Client.AwsStatusOk, ResultStatus.UPLOADED)]
+        [TestCase(SFS3Client.EXPIRED_TOKEN, ResultStatus.RENEW_TOKEN)]
+        [TestCase(MockS3Client.AwsStatusError, ResultStatus.NEED_RETRY)] // Any error that isn't the above will return ResultStatus.NEED_RETRY
+        public void TestUploadFile(string requestKey, ResultStatus expectedResultStatus)
+        {
+            // Arrange
+            var mockAmazonS3Client = new Mock<AmazonS3Client>(AwsKeyId, AwsSecretKey, AwsToken, _clientConfig);
+            mockAmazonS3Client.Setup(client => client.PutObjectAsync(It.IsAny<PutObjectRequest>(), It.IsAny<CancellationToken>()))
+                .Returns<PutObjectRequest, CancellationToken>((request, cancellationToken) =>
+                {
+                    if (request.BucketName == HttpStatusCode.OK.ToString())
```

Comment:
similarly we could think of extracting Put response to a separate method

Reply:
Sure